### PR TITLE
Feat/issue 3239 allow user expressions

### PR DESCRIPTION
**Pull request by iamonuwa**
##### Description

Allow users update status on Gitcoin

##### Refers/Fixes

Fixes #3239

##### Testing

To test this, go to your profile and add a post, it will display under the activity feed.

### CODE REVIEW

**Comment by iamonuwa** [comment]
@octavioamu @danlipert @thelostone-mc @owocki where will be the best place to place the status box?

**Comment by iamonuwa** [comment]
Link to the video clip for this PR https://www.dropbox.com/s/iczj46bov2zx4xp/Status%20Update.mp4?dl=0

**Comment by owocki** [comment]
> @octavioamu @danlipert @thelostone-mc @owocki where will be the best place to place the status box?

right above the activity feed.    if/when we merge https://github.com/gitcoinco/web/pull/4818 thatll put it towards the top

File: app/assets/v2/js/status.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    );
+    fetch('/api/v0.1/activity', {
+      method: 'post',
+      body: data
+    })
+      .then(response => {
+        if (response.status === 200) {
+          _alert(
+            { message: gettext('Status has been saved.') },
+            'success'
```
Comment (owocki) [inline app/assets/v2/js/status.js]: youll porbably want to paint it in the activity portion of the profile upon success too.

File: app/bounty_requests/templates/bounty_request_form.html [hunk 2]
```diff
@@ -27,24 +27,53 @@
...
-                  {% endif %}
-                </label>
-                <input name="requested_by" placeholder="{% trans "Please Login First" %}" id="requested_by" class="form__input" type="text" value="{% if github_handle %}{{ github_handle }}{% endif %}" required disabled />
+      <div class="col body">
+        <div class="container-fluid no-gutter">
+          <div class="row">
+            <div class="col-12 offset-md-1 col-md-10">
+              <div class="m-4">
+                <div>
+                  <div class="request-form" id="primary_form">
```
Comment (owocki) [inline app/bounty_requests/templates/bounty_request_form.html]: this seems unrelated?

File: app/dashboard/templates/profiles/profile.html [hunk 3]
```diff
@@ -204,6 +204,12 @@ <h1 class="profile-header__handle">
         </div>
       {% endif%}
 
+      <div class="container px-md-0">
+        <div class="row">
```
Comment (owocki) [inline app/dashboard/templates/profiles/profile.html]: only show this IFF it's the users own profile

File: app/retail/views.py [hunk 4]
```diff
@@ -1001,6 +1001,27 @@ def activity(request):
...
+    if request.POST:
+        profile = request.user.profile
+        kwargs = {
+            'activity_type': 'status_update',
+            'metadata': {
+                'title': request.POST.get('data')
+            }
+        }
+        kwargs['profile'] = profile
+        try:
```
Comment (owocki) [inline app/retail/views.py]: rate limit pls. 

also we'll need to make sure ppl don't try to XSS here.

**Comment by owocki** [comment]
refers link is off

File: app/dashboard/templates/profiles/profile.html (hunk 3 above)
Comment (iamonuwa) [reply app/dashboard/templates/profiles/profile.html]: 🤔

File: app/retail/views.py (hunk 4 above)
Comment (iamonuwa) [reply app/retail/views.py]: CSRF is in place and there is rate limit decorator on top of the methid

File: app/bounty_requests/templates/bounty_request_form.html (hunk 2 above)
Comment (iamonuwa) [reply app/bounty_requests/templates/bounty_request_form.html]: Rebasing with upstream master, introduced this

**Comment by iamonuwa** [comment]
> refers link is off

Can you elaborate please?

File: app/bounty_requests/templates/bounty_request_form.html (hunk 2 above)
Comment (thelostone-mc) [reply app/bounty_requests/templates/bounty_request_form.html]: @iamonuwa could redo this PR to ensure only your changes are present ?

**Comment by owocki** [comment]
i dont know what i was saying.. disregard..

**Comment by iamonuwa** [comment]
Cool

**Comment by iamonuwa** [comment]
@owocki @danlipert @thelostone-mc reviews are required here

**Comment by owocki** [comment]
this LGTM (left 1 comment)... i if its good with the engineers we can merge / payout. ive folded it into my profile workstream

File: app/assets/v2/js/status.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      $('#status input[name="csrfmiddlewaretoken"]').attr('value')
+    );
+    fetch('/api/v0.1/activity', {
+      method: 'post',
+      body: data
+    })
+      .then(response => {
+        if (response.status === 200) {
+          _alert(
+            { message: gettext('Status has been saved.') },
```
Comment (owocki) [inline app/assets/v2/js/status.js]: would it be possible to show the newly posted status update on the profile without a refresh??

File: app/retail/views.py [hunk 6]
```diff
@@ -994,6 +994,27 @@ def activity(request):
 
     return TemplateResponse(request, 'activity.html', context)
 
+@ratelimit(key='ip', rate='30/m', method=ratelimit.UNSAFE, block=True)
```
Comment (owocki) [inline app/retail/views.py]: 2/m seems more appopriate

File: app/assets/v2/js/status.js (hunk 5 above)
Comment (iamonuwa) [reply app/assets/v2/js/status.js]: To me, looks like overkill. Not sure what you think.

Comment (iamonuwa) [reply app/assets/v2/js/status.js]: As the other activity logs are not updating in realtime as well

**Comment by danlipert** [comment]
@iamonuwa Which issue is this PR referring to? Looks like you linked the wrong issue in the PR description

File: app/retail/views.py [hunk 7]
```diff
@@ -994,6 +994,27 @@ def activity(request):
...
+    response = {}
+    if request.POST:
+        profile = request.user.profile
+        kwargs = {
+            'activity_type': 'status_update',
+            'metadata': {
+                'title': request.POST.get('data')
+            }
+        }
+        kwargs['profile'] = profile
```
Comment (danlipert) [inline app/retail/views.py]: You can combine this line with the lines above it to create the `kwargs` dictionary in one statement

File: app/retail/views.py [hunk 8]
```diff
@@ -994,6 +994,27 @@ def activity(request):
...
+        try:
+            Activity.objects.create(**kwargs)
+            response['status'] = 200
+            response['message'] = 'Status updated!'
+        except Exception as e:
+            response['status'] = 400
+            response['message'] = 'Bad Request'
+            logger.error('Status Update error - Error: (%s) - Handle: (%s)', e, profile.handle if profile else '')
```
Comment (danlipert) [inline app/retail/views.py]: Instead of constructing the response objects line-by-line, lets just return a dictionary directly

File: app/retail/views.py [hunk 9]
```diff
@@ -994,6 +994,27 @@ def activity(request):
...
+            'metadata': {
+                'title': request.POST.get('data')
+            }
+        }
+        kwargs['profile'] = profile
+        try:
+            Activity.objects.create(**kwargs)
+            response['status'] = 200
+            response['message'] = 'Status updated!'
+        except Exception as e:
```
Comment (danlipert) [inline app/retail/views.py]: What exceptions are we planning on handling here?

Comment (iamonuwa) [reply app/retail/views.py]: Come in handy for catching and logging errors

Comment (danlipert) [reply app/retail/views.py]: Yeah but is there a specific exception we are planning on seeing? We could put try: catch blocks around everything but catching exceptions costs performance, so we need to use it only in cases where we need it.

Comment (iamonuwa) [reply app/retail/views.py]: Should I remove it?

**Comment by thelostone-mc** [comment]
status updates are available in the new profile, so we can close this out ^_^
